app/controllers/api/v1/materials_controller.rb

```diff
@@ -44,6 +44,264 @@ def featured
+                                            favoritable_type: rubyclass.name )
+                if favorites.count == 1
+
+                  favorite = favorites[0] 
+                  message = "Favorite already exists with id #{favorite.id}."
+                  favorite_id = favorite.id
+
+                else
+                  #
+                  # This shouldn't happen...
+                  #
+                  status = 400
+                  message = "Could not create favorite even though " +
+                            "#{favorites.count} favorite(s) already exist."
+                end
+              end
+
+            else 
+              status = 400
+              message = "Invalid material id #{id}"
+            end
+
+          else
+            status = 400
+            message = "Invalid material type #{type}"
+          end
+
+        rescue ActiveRecord::RecordNotFound => rnf
+          status = 400
+          message = "RecordNotFound Invalid material id #{id}"
+        end
+
+      else
+        status = 400
+        message = "Missing material type (#{type}) or id (#{id})"
+      end
+
+    else
+      status = 400
+      message = "Cannot add favorite for non-logged in user."
+    end
+
+    render json: {  :message        => message, 
+                    :favorite_id    => favorite_id  }, :status => status
+
+  end
+
+  #
+  # Get all favorites for the currently logged in user.
+  #
+  def get_favorites
+
+    status  = 200
+    data    = nil
+
+    if current_user && !current_user.anonymous?
+
+      favorites     = current_user.favorites
+      type_ids_map  = {}
+      materials     = []
+
+      #
+      # Build sets of IDs for each type
+      #
+      favorites.each do |favorite|
+        favoritable_type    = favorite.favoritable_type
+        favoritable_id      = favorite.favoritable_id
+        if !type_ids_map[favoritable_type] 
+          type_ids_map[favoritable_type] = []
+        end
+        type_ids_map[favoritable_type].append(favoritable_id)
+      end
+
+      #
+      # Optimize queries for different material types.
+      #
+      includes_map = {}
+
+      includes_map['ExternalActivity']  = [ :template,
+                                            :user,
+                                            :subject_areas,
+                                            :grade_levels       ]
+
+      includes_map['Interactive']       = [ :user,
+                                            :subject_areas,
+                                            :grade_levels       ]
+
+      includes_map['Investigation']     = [ ]
+
+      #
+      # Now do a single query for each type
+      #
+      type_ids_map.each do |key, val| 
+        list = includes_map[key]
+        items = key.constantize.includes(*list).find(val)   
+                                # E.g. the above might become:
+                                # ExternalActivity.includes(...).find([1,2,3])
+        materials += items
+      end
```

Comment:
I'd recommend unrolling this this loop. Then you won't need to define the includes_map up above. It will be fewer lines of code overall and will be easier to read.

So it would become: 
```
materials += ExternalActivity.includes(....).find(type_ids_map['ExternalActivity'])
materials += Interactive.includes(...).find(type_ids_map['Interactive'])